File: old/core/aggregators/registry.py
"""Aggregator registry for mapping feed types to classes."""

from typing import Dict, Type

from .ars_technica import ArsTechnicaAggregator
from .base import BaseAggregator
from .caschys_blog.aggregator import CaschysBlogAggregator
from .dark_legacy.aggregator import DarkLegacyAggregator
from .explosm.aggregator import ExplosmAggregator
from .heise import HeiseAggregator
from .implementations import FeedContentAggregator
from .mactechnews.aggregator import MactechnewsAggregator
from .mein_mmo import MeinMmoAggregator
from .merkur import MerkurAggregator
from .oglaf import OglafAggregator
from .podcast.aggregator import PodcastAggregator
from .reddit import RedditAggregator
from .tagesschau import TagesschauAggregator
from .the_verge import TheVergeAggregator
from .website import FullWebsiteAggregator
from .youtube.aggregator import YouTubeAggregator


class AggregatorRegistry:
    """Registry for aggregator classes."""

    _registry: Dict[str, Type[BaseAggregator]] = {
        "full_website": FullWebsiteAggregator,
        "feed_content": FeedContentAggregator,
        "heise": HeiseAggregator,
        "merkur": MerkurAggregator,
        "tagesschau": TagesschauAggregator,
        "explosm": ExplosmAggregator,
        "dark_legacy": DarkLegacyAggregator,
        "caschys_blog": CaschysBlogAggregator,
        "mactechnews": MactechnewsAggregator,
        "oglaf": OglafAggregator,
        "mein_mmo": MeinMmoAggregator,
        "the_verge": TheVergeAggregator,
        "ars_technica": ArsTechnicaAggregator,
        "youtube": YouTubeAggregator,
        "reddit": RedditAggregator,
        "podcast": PodcastAggregator,
    }

    @classmethod
    def get(cls, aggregator_type: str) -> Type[BaseAggregator]:
        """
        Get aggregator class for the given type.

        Args:
            aggregator_type: The aggregator type string (e.g., 'full_website')

        Returns:
            Aggregator class

        Raises:
            KeyError: If aggregator type is not found
        """
        if aggregator_type not in cls._registry:
            raise KeyError(f"Unknown aggregator type: {aggregator_type}")
        return cls._registry[aggregator_type]

    @classmethod
    def get_all(cls) -> Dict[str, Type[BaseAggregator]]:
        """Get all registered aggregators."""
        return cls._registry.copy()


def get_aggregator(feed) -> BaseAggregator:
    """
    Get aggregator instance for a feed.

    Args:
        feed: Feed model instance

    Returns:
        Instantiated aggregator
    """
    aggregator_class = AggregatorRegistry.get(feed.aggregator)
    return aggregator_class(feed)

File: src/lib/aggregators/registry.ts
/**
 * SERVER-ONLY: every class below reaches `@/lib/db/client` (and therefore
 * `better-sqlite3`) through the image store. The option *descriptions* a form
 * needs live in `./specs`, which imports none of this — see the comment there.
 *
 * `IMPLEMENTED_AGGREGATORS` is the one map from an `AggregatorKey` to its
 * class. `./factory`'s `createAggregator()` is the only thing that reads it
 * to instantiate an aggregator -- there used to be a second entry point here
 * too (`AggregatorRegistry.get`/`getAll` + `getAggregator()`), but nothing
 * called it outside this module's own tests, and it disagreed with
 * `createAggregator()` on an unknown key (`.get` threw; `createAggregator`
 * falls back to `FullWebsiteAggregator`, which is the behaviour every real
 * caller depends on). Removed rather than reconciled, per the 2026-09-03
 * pipeline-review-4 cleanup plan's Task 2.
 */
import type { FeedLike } from "./base";
import { BaseAggregator } from "./base";
import { RssAggregator } from "./rss";
import { ArsTechnicaAggregator } from "./sites/ars_technica";
import { CaschysBlogAggregator } from "./sites/caschys_blog";
import { DarkLegacyAggregator } from "./sites/dark_legacy";
import { ExplosmAggregator } from "./sites/explosm";
import { HeiseAggregator } from "./sites/heise";
import { MactechnewsAggregator } from "./sites/mactechnews/aggregator";
import { MeinMmoAggregator } from "./sites/mein_mmo/aggregator";
import { MerkurAggregator } from "./sites/merkur";
import { OglafAggregator } from "./sites/oglaf";
import { PodcastAggregator } from "./sites/podcast";
import { RedditAggregator } from "./sites/reddit/aggregator";
import { TagesschauAggregator } from "./sites/tagesschau/aggregator";
import { TheVergeAggregator } from "./sites/the_verge";
import { YouTubeAggregator } from "./sites/youtube/aggregator";
import { FullWebsiteAggregator } from "./website";

export type AggregatorClass = new (feed: FeedLike) => BaseAggregator;

export const IMPLEMENTED_AGGREGATORS: Record<string, AggregatorClass | undefined> = {
  feed_content: RssAggregator as unknown as AggregatorClass,
  rss: RssAggregator as unknown as AggregatorClass,
  full_website: FullWebsiteAggregator as unknown as AggregatorClass,
  oglaf: OglafAggregator as unknown as AggregatorClass,
  dark_legacy: DarkLegacyAggregator as unknown as AggregatorClass,
  explosm: ExplosmAggregator as unknown as AggregatorClass,
  caschys_blog: CaschysBlogAggregator as unknown as AggregatorClass,
  merkur: MerkurAggregator as unknown as AggregatorClass,
  the_verge: TheVergeAggregator as unknown as AggregatorClass,
  ars_technica: ArsTechnicaAggregator as unknown as AggregatorClass,
  mactechnews: MactechnewsAggregator as unknown as AggregatorClass,
  tagesschau: TagesschauAggregator as unknown as AggregatorClass,
  heise: HeiseAggregator as unknown as AggregatorClass,
  mein_mmo: MeinMmoAggregator as unknown as AggregatorClass,
  podcast: PodcastAggregator as unknown as AggregatorClass,
  youtube: YouTubeAggregator as unknown as AggregatorClass,
  reddit: RedditAggregator as unknown as AggregatorClass,
};
